ui/raidboss/emulator/data/LogEventHandler.ts

```diff
@@ -0,0 +1,119 @@
+import EmulatorCommon from '../EmulatorCommon';
+import EventBus from '../EventBus';
+import { LocaleNetRegex } from '../../../../resources/translations';
+import { Lang } from 'types/global';
+import LineEvent from './network_log_converter/LineEvent';
+import { LineEvent0x01 } from './network_log_converter/LineEvent0x01';
+
+export default class LogEventHandler extends EventBus {
+  static doesLineMatch(line: string, regexes: Record<string, RegExp>): RegExpExecArray | false {
+    for (const i in regexes) {
+      const res = regexes[i]?.exec(line);
+      if (res) {
+        if (LocaleNetRegex.areaSeal[i as Lang] && res.groups)
+          res.groups.language = i;
+        return res;
+      }
+    }
+    return false;
+  }
+
+  static isMatchStart(line: string): RegExpMatchArray | false {
+    let res;
+    res = LogEventHandler.doesLineMatch(line, EmulatorCommon.countdownRegexes);
+    if (res) {
+      res.groups = res.groups ?? {};
+      res.groups.StartIn = (parseInt(res.groups.time ?? '0') * 1000).toString();
+      res.groups.StartType = 'Countdown';
+      return res;
+    }
+    res = LogEventHandler.doesLineMatch(line, EmulatorCommon.sealRegexes);
+    if (res) {
+      res.groups = res.groups ?? {};
+      res.groups.StartIn = '0';
+      res.groups.StartType = 'Seal';
+      return res;
+    }
+    res = LogEventHandler.doesLineMatch(line, EmulatorCommon.engageRegexes);
+    if (res) {
+      res.groups = res.groups ?? {};
+      res.groups.StartIn = '0';
+      res.groups.StartType = 'Engage';
+      return res;
+    }
+    return false;
+  }
+
+  static isMatchEnd(line: string): RegExpMatchArray | false {
+    let res;
+    res = LogEventHandler.doesLineMatch(line, { '': EmulatorCommon.winRegex });
+    if (res) {
+      res.groups = res.groups ?? {};
+      res.groups.EndType = 'Win';
+      return res;
+    }
+    res = LogEventHandler.doesLineMatch(line, { '': EmulatorCommon.wipeRegex });
+    if (res) {
+      res.groups = res.groups ?? {};
+      res.groups.EndType = 'Wipe';
+      return res;
+    }
+    res = LogEventHandler.doesLineMatch(line, { '': EmulatorCommon.cactbotWipeRegex });
+    if (res) {
+      res.groups = res.groups ?? {};
+      res.groups.EndType = 'Cactbot Wipe';
+      return res;
+    }
+    res = LogEventHandler.doesLineMatch(line, EmulatorCommon.unsealRegexes);
+    if (res) {
+      res.groups = res.groups ?? {};
+      res.groups.EndType = 'Unseal';
+      return res;
+    }
+    return false;
+  }
+
+  public currentFight: LineEvent[] = [];
+  public currentZoneName?: string;
+  public currentZoneId = -1;
```

Comment:
Unrelated to this patch, but do you think we need a `ZoneId.InvalidZone`?

Reply:
I don't think that having a `ZoneId.InvalidZone` makes sense because every zone ID should always be valid. It's more of a case of `ZoneId.UnknownZone` as a catch-all which I'm not sure is really possible?